File: src/id0068.c
// Licensed under the MIT License.

// Magic 5-gon Ring

#include <math.h>
#include "../lib/euler.h"
#include "../lib/permutation_iterator.h"

int main(void)
{
    int l[] = { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 };
    unsigned long long max = 0;
    struct PermutationIterator it;
    clock_t start = clock();
    
    for (permutation_begin(&it, l, sizeof * l, 10, int_comparer);
        !it.end;
        permutation_next(&it))
    {
        if (l[0] > l[3] || l[0] > l[5] || l[0] > l[7] || l[0] > l[9] ||
            (l[3] != 10 && l[5] != 10 && l[7] != 10 && l[9] != 10))
        {
            continue;
        }

        int sum = l[0] + l[1] + l[2];

        if (l[2] + l[3] + l[4] != sum || l[4] + l[5] + l[6] != sum ||
            l[6] + l[7] + l[8] != sum || l[8] + l[9] + l[1] != sum)
        {
            continue;
        }
        
        unsigned long long value = math_concat(
            l[0], l[1], l[2], 
            l[3], l[2], l[4], 
            l[5], l[5], l[6],
            l[7], l[6], l[8],
            l[9], l[8], l[1]);

        if (value > max)
        {
            max = value;
        }
    }

    return euler_submit(68, max, start);
}

File: src/id0068.py
# Licensed under the MIT License.

# Magic 5-gon Ring

from itertools import permutations
from time import time

maxString = ""
start = time()

for a, b, c, d, e, f, g, h, i, j in permutations(range(1, 10 + 1)):
    if a > d or a > f or a > h or a > j: continue
    if d != 10 and f != 10 and h != 10 and j != 10: continue
    
    sum = a + b + c
    
    if c + d + e != sum or e + f + g != sum: continue
    if g + h + i != sum or i + j + b != sum: continue
    
    string = "".join(map(str, [ a, b, c, d, c, e, f, e, g, h, g, i, j, i, b ]))
    
    if string > maxString:
        maxString = string

print(f"0068{maxString:>64}    {time() - start:.6f}")
